src/coreclr/src/vm/nativeimage.cpp

```diff
@@ -0,0 +1,156 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+// --------------------------------------------------------------------------------
+// NativeImage.cpp
+//
+
+// --------------------------------------------------------------------------------
+
+#include "common.h"
+#include "nativeimage.h"
+
+// --------------------------------------------------------------------------------
+// Headers
+// --------------------------------------------------------------------------------
+
+#include <shlwapi.h>
+
+BOOL AssemblyNameIndexHashTraits::Equals(LPCUTF8 a, LPCUTF8 b)
+{
+    WRAPPER_NO_CONTRACT;
+
+    SString s1;
+    SString s2;
+    s1.SetUTF8(a);
+    s2.SetUTF8(b);
+    return s1.CompareCaseInsensitive(s2) == 0;
+}
+
+AssemblyNameIndexHashTraits::count_t AssemblyNameIndexHashTraits::Hash(LPCUTF8 s)
+{
+    WRAPPER_NO_CONTRACT;
+
+    SString s1;
+    s1.SetUTF8(s);
+    return s1.HashCaseInsensitive();
+}
+
+NativeImage::NativeImage(
+    NewHolder<PEImageLayout>& peImageLayout,
```

Comment:
It is unusual to pass NewHolder by reference like this in CoreCLR codebase. The usual pattern is to have a constructor that takes ownership and cannot fail half-way through, and Initialize method that does the initialization that can fail in arbitrary ways.

Reply:
Fixed in 16th commit.